File: app/src/main/java/com/foxek/simpletimer/ui/workout/WorkoutAdapter.java
package com.foxek.simpletimer.ui.workout;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.foxek.simpletimer.R;
import com.foxek.simpletimer.data.model.Workout;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class WorkoutAdapter extends ListAdapter <Workout, WorkoutAdapter.ViewHolder> {

    private Context             mContext;
    private Callback            mCallback;

    @Inject
    public WorkoutAdapter() {
        super(DIFF_CALLBACK);
    }

    public interface Callback {
        void onListItemClick(Workout workout);
    }

    void setCallback(Callback callback) {
        mCallback = callback;
    }

    @Override
    public void onBindViewHolder(@NonNull WorkoutAdapter.ViewHolder holder, int position) {
        holder.mWorkoutName.setText(getItem(position).getName());
        holder.mWorkoutDescription.setText(mContext.getString(R.string.number_of_intervals_text,
                getItem(position).getIntervalCount()));
    }

    @NonNull
    @Override
    public WorkoutAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.workout_item, parent, false);
        return new WorkoutAdapter.ViewHolder(view);
    }

    private static final DiffUtil.ItemCallback<Workout> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<Workout>() {
                @Override
                public boolean areItemsTheSame(
                        @NonNull Workout oldWorkout, @NonNull Workout newWorkout) {
                    return ((oldWorkout.getName().equals(newWorkout.getName())) &&
                            (oldWorkout.getIntervalCount() == newWorkout.getIntervalCount()) &&
                            (oldWorkout.getUid() == newWorkout.getUid()));
                }
                @Override
                public boolean areContentsTheSame(
                        @NonNull Workout oldWorkout, @NonNull Workout newWorkout) {
                    return ((oldWorkout.getName().equals(newWorkout.getName())) &&
                            (oldWorkout.getIntervalCount() == newWorkout.getIntervalCount()) &&
                            (oldWorkout.getUid() == newWorkout.getUid()));
                }
            };

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.workout_name)
        TextView mWorkoutName;

        @BindView(R.id.workout_description)
        TextView mWorkoutDescription;

        ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
            mContext = v.getContext();
        }

        @OnClick({R.id.workout_item})
        @Override
        public void onClick(View v) {
            mCallback.onListItemClick(getItem(getAdapterPosition()));
        }
    }
}

File: app/src/main/java/com/foxek/simpletimer/ui/workout/adapter/WorkoutAdapter.kt
package com.foxek.simpletimer.ui.workout.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.foxek.simpletimer.R

import javax.inject.Inject
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.foxek.simpletimer.data.model.Workout

import kotlinx.android.synthetic.main.workout_item.view.*

class WorkoutAdapter @Inject constructor() : ListAdapter<Workout, WorkoutAdapter.ViewHolder>(WorkoutDiffCallback()) {

    private var mCallback: Callback? = null

    interface Callback {
        fun onListItemClick(workout: Workout)
    }

    internal fun setCallback(callback: Callback) {
        mCallback = callback
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.itemView.item_workout_name.text = getItem(position).name
        holder.itemView.item_workout_description.text = holder.itemView.resources.getString(R.string.number_of_intervals_text,
                getItem(position).intervalCount)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.workout_item, parent, false)
        return ViewHolder(view)
    }

    inner class ViewHolder constructor(v: View) : RecyclerView.ViewHolder(v) {

        init {
            itemView.workoutItem.setOnClickListener {
                mCallback?.onListItemClick(getItem(adapterPosition))
            }
        }
    }
}
